### ISSUE TITLE
Mirror the manual on the web site?

### ISSUE THREAD

**Issue by pmetzger** [ocaml/v2.ocaml.org#1011]
The only time I ever go to caml.inria.fr is to look at the manual online. I know there's a thought of eventually having "docs.ocaml.org" or some such, but in the meanwhile, would it be a problem to mirror the manual from http://caml.inria.fr/pub/docs/manual-ocaml/ to some URL on ocaml.org? I would submit a PR, but the build system for the web site is a bit opaque.

**Comment by Chris00** [comment]
The problem is not the technical details, it is to have the agreement of Inria because we were specifically asked not to mirror files in `pub/`.

**Comment by Chris00** [comment]
Maybe one thing we could do is to put in place some "proxying" of the URLs so they seem to be located on o.o while actually being hosted on Inria servers.  @xavierleroy or @damiendoligez may say whether this is a satisfactory solution for them.

**Comment by pmetzger** [comment]
Isn't the manual redistributable?

**Comment by agarwal** [comment]
My understanding was that the Inria team wanted to keep organizing files under `pub/`, i.e. they didn't want their own workflow altered. I don't believe there is any opposition to mirroring content elsewhere if someone else does that work.

**Comment by pmetzger** [comment]
We could also just build the 4.07 manual ourselves. It's straightforward.

**Comment by xavierleroy** [comment]
> The problem is not the technical details, it is to have the agreement of Inria because we were specifically asked not to mirror files in pub/.

I forgot this discussion.  At any rate things have changed since the early design days of ocaml.org: for example, the preferred URL for source distributions is now https://github.com/ocaml/ocaml/releases/ instead of http://caml.inria.fr/pub/, because the former is more reliable.  More generally, I have concerns about the long-term availability of caml.inria.fr/

So, I have no problem with hosting copies of the OCaml manuals on ocaml.org, and it can only increase availability of the material.  Bonus points if you can have the manual automatically generated from source distributions or even from Github branches.

**Comment by pmetzger** [comment]
> Bonus points if you can have the manual automatically generated from source distributions or even from Github branches.

This should be quite doable. Scripting git is quite easy, and the github API is easy to use, too. I could work on it if I had some sense of how to run things that generate HTML for ocaml.org.

**Comment by pmetzger** [comment]
BTW, if someone who can explain to me a bit more about the principles of how the ocaml.org build system is organized would get in touch with me, I'll do a pull request to do this.

**Comment by agarwal** [comment]
@pmetzger Might be easier to exchange some emails about this. I'll get in touch.

**Comment by Chris00** [comment]
Any news on this?

**Comment by pmetzger** [comment]
Nope, I dropped it on the floor. I'm a bit busy for the next couple of weeks, so if someone else wants to do it meanwhile I won't object. Otherwise keep poking me.

**Comment by sanette** [comment]
I have started working on this. I need help on what the rules are for integrating new pages

**Comment by sanette** [comment]
Next question, is _where_ do we store the manual on `ocaml.org`?
A possibility would be to modify the `https://ocaml.org/docs/` so that the first 4 "blocks" would be
links to the 4 parts of the manual. If you agree with this, I can try to propose something

**Comment by pmetzger** [comment]
I'd suggest `https://ocaml.org/docs/manual-ocaml/` as a URL, and have everything on `https://ocaml.org/docs/` that points currently to the manual at caml.inria.fr point instead there. But this does mean that we need an easy procedure for updating the manual when OCaml gets updated!

**Comment by dbuenzli** [comment]
> I'd suggest `https://ocaml.org/docs/manual-ocaml/` as a URL,

I know `manual-ocaml` is what is on caml.inria.fr but I have a slight gripe with `manual-ocaml` in that the opam package for the manual is called `ocaml-manual` (my fault). 

What about simply `https://ocaml.org/manual`, that's easy to type and remember.

**Comment by sanette** [comment]
> But this does mean that we need an easy procedure for updating the manual when OCaml gets updated!

yes of course. At this point I have a script that does everything but only for Part 1 of the manual, and I have only tested it for the latest version of ocaml. What I could do when I get the time is to test it against all available versions of ocaml to see if this is robust.

**Comment by Chris00** [comment]
As the URL, I suggest `https://ocaml.org/docs/manual` as a link pointing to the most recent manual and `https://ocaml.org/docs/manual/⟨version⟩` for the actual manual for OCaml ⟨version⟩.  So it would be ideal for the script to be passed a ⟨version⟩ and build the manual from the corresponding Git branch.  The main Makefile already has a variable for the most recent OCaml version so can adequately drive this script.

**Comment by pmetzger** [comment]
@dbuenzli suggests:
> What about simply `https://ocaml.org/manual`, that's easy to type and remember.

Seems fine by me, as does @Chris00's suggestion of `https://ocaml.org/docs/manual` — @dbuenzli's suggestion has the advantage of being slightly more compact. I agree that the `manual/[version]` idea would also be a good idea to allow access to a specific version.

**Comment by dbuenzli** [comment]
I'd prefer  `/manual`. That way I don't need a search engine or try to do remember whether it's `doc` or `docs`. +1 for versions.

**Comment by Chris00** [comment]
I do not have strong opinions about `/docs/manual` or `/manual`.  For docs v.s. doc, maybe we should add a rewrite rule doc → docs (on the server) so it is not an issue?

**Comment by dbuenzli** [comment]
> For docs v.s. doc, maybe we should add a rewrite rule doc → docs (on the server) so it is not an issue?

I don't think that's needed if I don't need to remember it or type it manually.

**Comment by pmetzger** [comment]
I'd say the consensus is now `https://ocaml.org/manual` for the latest, and `https://ocaml.org/manual/[version]` for a particular version of the manual.

**Comment by sanette** [comment]
The script is now able to process versions 4.00 to 4.09. See:
https://sanette.github.io/ocaml-tutorial/

**Comment by agarwal** [comment]
I prefer `/manual` over `/docs/manual`. This follows the principle of "flat is better than nested". Unless we anticipate a name clash, but I think "manual" uniquely refers to the single document we are talking about here.

> For docs v.s. doc, maybe we should add a rewrite rule doc → docs (on the server) so it is not an issue?

I think that's not worth it. This introduces two names for the same thing, which I think is generally not a good idea.

**Comment by sanette** [comment]
Is there a need to include versions older than 4.00?

**Comment by pmetzger** [comment]
@sanette I don't think so. Indeed, I think we can probably cut it off at 4.02 but you've already gone earlier than that.

**Comment by sanette** [comment]
Hi, I need some help... I'm generating the ocaml.org website locally, and trying to add the manual pages. This is driving me nuts. Here is an exemple: I modify ocaml.org/site/docs/index.md to
```
<!-- ((! set title Docs !)) ((! set documentation !)) ((! set nobreadcrumb !)) -->

<h1>Documentation</h1>
<pre><div class="hello">Hello</div></pre>
Bye
```
and then run `make`, and here is the generated code:
```html
<h1>Documentation</h1>
<pre></pre><div class="hello">Hello</div>

<p>Bye</p>
```
Notice how the `<pre>` has been closed prematurely.
What should I do??

**Comment by pmetzger** [comment]
That's a new one on me. I don't know the answer. I think you're going to have to debug the generation toolchain. :(

**Comment by sanette** [comment]
And I don't understand the logic of using markdown files to include HTML. This is very dangerous. For instance in the example above, if you inadvertently add 4 spaces before `<h1>` then it is recognized as code and it generates:
```html
<pre><code>&lt;h1&gt;Documentation&lt;/h1&gt;</code></pre>
```

**Comment by pmetzger** [comment]
I'm not a giant fan of the way that the web site is set up (it's too complicated, IMHO) but fixing that would need to be a separate PR. :)

**Comment by sanette** [comment]
> That's a new one on me. I don't know the answer. I think you're going to have to debug the generation toolchain. :(

unfortunately, this is a blocker for me... because the ocaml manual uses `<pre><div ...>` for essentially all code snippets... :(

**Comment by sanette** [comment]
![Screenshot_20191109_203156](https://user-images.githubusercontent.com/7536536/68534014-260e3400-0330-11ea-825b-8d5dbbbf5959.png)

Here is my current experimentation.

You can see at the very bottom that ocaml code is not correctly recognized by the CSS, because of this `<pre>` issue.

**Comment by Chris00** [comment]
I do not have time before Monday to look at the particular markdown problem you mention but you can output HTML files — they are also accepted.

**Comment by sanette** [comment]
@Chris00 , the problem is the same with HTML files, apparently they get modified.

**Comment by Chris00** [comment]
Not much, see [Makefile.from_html](https://github.com/ocaml/ocaml.org/blob/master/Makefile.from_html).  This also explains why most HTML code is in markdown files: to use [template/main.mpp](https://github.com/ocaml/ocaml.org/blob/master/template/main.mpp) to “surround” the content.

**Comment by sanette** [comment]
What I meant is that the particular problem with `<pre>` happens also with html files.
If I write this `index.html`:
```html
<h1>Documentation</h1>
<pre><div class="hello">Hello</div></pre>
Bye
```
then the generated `index.html` is
```html
<!DOCTYPE HTML>


<h1>Documentation</h1>
<pre></pre><div class="hello">Hello</div>
Bye
```

**Comment by sanette** [comment]
> I'd prefer `/manual`. That way I don't need a search engine or try to do remember whether it's `doc` or `docs`. +1 for versions.

Just a silly remark, but I lost one hour because of this: it turns out my ubuntu has a default option that aliases  `/manual` to the apache manual... 8-)

**Comment by sanette** [comment]
I have narrowed the problem down. At the end of the make process, the `script/relative_urls` is applied to all files (which is a bit brutal, by the way). It is this script that scrambles the `<pre>` tag away...

You can try it directly
```
script/relative_urls foo.html
```
where foo.html contains a `<pre><div...>` a above

**Comment by sanette** [comment]
So apparently the problem comes from the `Nethtml` module. Which library is this?

**Comment by Chris00** [comment]
Yes, I came to the same conclusion.  The script `script/relative_urls.ml` parses HTML for the local rendering so that links can be browsed on the disk.  The HTML is parsed with `Nethtml`:
```
# #require "netstring";;
# let html = 
  let s = {|<h1>Documentation</h1>
<pre><div class="hello">Hello</div></pre>
Bye|} in
  Nethtml.parse_document (Lexing.from_string s);;
val html : Nethtml.document list =
  [Nethtml.Element ("h1", [], [Nethtml.Data "Documentation"]);
   Nethtml.Data "\n"; Nethtml.Element ("pre", [], []);
   Nethtml.Element ("div", [("class", "hello")], [Nethtml.Data "Hello"]);
   Nethtml.Data "\nBye"]
```
I believe `Nethtml` rewrites the HTML that way because `<div>` tags are not allowed inside `<pre>`.

**Comment by sanette** [comment]
> `<div>` tags are not allowed inside `<pre>` 

 ==> in this case we have to tell this to hevea.
Since version 4.05, this is the way Hevea outputs ocaml code... :(

**Comment by Chris00** [comment]
[Issue opened](https://github.com/ocaml/ocaml/issues/9109).  This is not hevea's fault.

**Comment by sanette** [comment]
thanks

**Comment by sanette** [comment]
In the meantime I'm asking lambdasoup to replace these `div`s by `code`, so I can continue working on the manual.

**Comment by sanette** [comment]
Following our previous discussions, here is what I propose for the `/docs` page. 
![Screenshot_20191116_174911](https://user-images.githubusercontent.com/7536536/68996795-3e85cd80-089f-11ea-96e1-920120198069.png)

I will do a PR so that you can test and try by yourselves. But if you already have some comments, I can try to implement them. Of course it's not finished! At this point only the tutorial part links to the local version of the manual. The other parts link to pages on the Inria website. The 'tools' part is still missing, I will work on it when I can.

**Comment by sanette** [comment]
I'm not very satisfied with the distinction between "Core Library" and "Standard Library". Maybe we should put both in the same block. 
Btw in the manual itself it's not so clear: sometimes the Core Library (Stdlib) is called Standard Library, and sometimes all other modules are also called Standard Library.

**Comment by sanette** [comment]
I think this one is better:
![Screenshot_20191117_122256](https://user-images.githubusercontent.com/7536536/69006820-2bbdd800-0935-11ea-9b5e-73311a21b950.png)

**Comment by sanette** [comment]
you can check it out from my fork:

https://github.com/sanette/ocaml.org

**Comment by agarwal** [comment]
> the `script/relative_urls` is applied to all files (which is a bit brutal, by the way).

We should simply tell developers to run a dev server to view the site locally. A trivial `python -m SimpleHTTPServer` is all that is needed, and we could delete the URL munging.

> it turns out my ubuntu has a default option that aliases /manual to the apache manual

This would also be resolved.

> I'm not very satisfied with the distinction between "Core Library" and "Standard Library". 

The library that ships with OCaml is called the "Standard Library". The "Core Library" is not the right name for it. That is an established name, and it refers to Jane Street's Core suite of libraries. I know the manual has a section called the "The core library". The term isn't actually important to the library distributed with the compiler, and I personally think we should do away with the term "core" in that context. All of it is packaged in a module called `Stdlib` and we can call it all "The Standard Library".

**Comment by xavierleroy** [comment]
For what it's worth:
- What is called "the core library" in the OCaml manual is everything that is available by default (without `Module.` qualification) in the initial environment.  It comprises built-in definitions and the contents of the `Stdlib` module.  We thought it important to document these first, before the standard library modules (those that require `Module.` qualification or `open Module` statements).  If you can preserve that order of presentation in your online version of the manual, that's good enough.  What would NOT be good enough is to document the `Stdlib` module like the other modules, e.g. between `StdLabels` and `Stream` in alphabetic order of module names.
- The OCaml manual was using the terminology "The core library" well before Jane Street named their library "Core".  

More generally: I would really like to see this effort (of hosting the OCaml manual on ocaml.org) reach completion.  The future of the caml.inria.fr Web server is still unclear, and the manual deserves a more stable home.

**Comment by agarwal** [comment]
It's an unfortunate name clash. We can probably avoid any confusion easily though. For example, perhaps always say "The Stdlib core", with "core" in lower case. My concern was really "The Core Library" as a title, as shown in the image [above](https://github.com/ocaml/ocaml.org/issues/1011#issuecomment-554657852).

**Comment by pmetzger** [comment]
> More generally: I would really like to see this effort (of hosting the OCaml manual on ocaml.org) reach completion. The future of the caml.inria.fr Web server is still unclear, and the manual deserves a more stable home.

I think this can happen without much more work. At worst we can just put up the pages as is without any adaptation until we can come up with a process to make it a bit prettier.

**Comment by damiendoligez** [comment]
@pmetzger @sanette any news on this?

**Comment by sanette** [comment]
Hi Damien, personally I think that what I proposed on my fork (see message above) is quasi-ready to be put online, but I am not sure others would agree, as I had no feedback. 
My opinion is that we should do it, even if later we reconsider; for instance if we want to move to sphinx.

**Comment by pmetzger** [comment]
@sanette Can you do a pull request?

**Comment by sanette** [comment]
ok I can work on this. As an aside question, can someone look into why the 'make' script in the ocaml.org dir is so slow (and makes my computer heat up). Each time I make a small modif and re-generate the 300 files of the Stdref, 'make' takes forever to preprocess the MD files. (maybe 10min -- about 2sec per file). Compared to the 2-3 seconds necessary to generate the 300 MD files, it makes me wonder whether there is a bug in the Makefile.

**Comment by sanette** [comment]
apparently for each .md file processsed, an `opam` instance is run, and fetches something on internet. This probably contributes to the slowness.
I we want to host the full ocaml manual, I think we should find a way to circumvent this to have a fast "make" each time we update the manual.

**Comment by sanette** [comment]
I have uploaded all the API for 4.09 on my fork: https://github.com/sanette/ocaml.org
So we have now:

"tutorials" = chapters1--6" (imported from https://github.com/sanette/ocaml-tutorial)
"api" = chapter 25 (imported from https://github.com/sanette/ocaml-api)

It comes closer to something worth of a PR, however these are still missing:
* french translation of the doc page (and of all other manual/api pages, of course)
* API (chapter 25) for ocaml < 4.09
* all the "tools" pages (chapters 9-23), all the "language ref & extensions" (chapters 7-8) = I simply redirect to the Inria pages for these.
* chapters 27-32 (although the API part is already there)

**Comment by dbuenzli** [comment]
I'm not sure I understand fully what is being done here but I have a few concerns regarding [this screenshoot](https://github.com/ocaml/ocaml.org/issues/1011#issuecomment-554736257) and the end-user experience.

1. The way to access the manual for a particular version on that screenshot doesn't seem very user friendly. It's unlikely that I want to access these newly created subsets of the manual for a particular version individually, you likely want to consult the manual for a particular version altogether.
2. I don't think it's a good idea to specifically remix the manual for `ocaml.org` without doing that in the upstream manual. One thing that we have now is that you can get the same manual with the same structure that you have on `caml.inria.fr` locally by simply installing the `ocaml-manual` package. Remixing the manual only for `ocaml.org` loses that property. I think that changing the structure of the manual, if really needed, should be done upstream.
2. If you use `odig` to get documentation about your libraries `odig` links either on your local version of the manual (if you installed the `ocaml-manual` package) or the online version. For the reader there's no structural difference between the two and this is a good thing (there may be however theming difference depending on the `odig` theme which may provide a custom `manual.css`, see [here](https://erratique.ch/software/odig/doc/packaging.html#theme_install) for details). 

As far as I'm concerned the title of this issue is right: just mirror the OCaml manual on ocaml.org as published by upstream as follows: 
```
ocaml.org/manual -> http 307 to ocaml.org/manual/$(latest_version)
ocaml.org/manual/X.YY.ZZ -> ocaml manual for version X.YY.ZZ
```
We can also have a page somewhere in `/docs` that has the list of these url with appropriate text and descriptions. 

And if the manual needs structural changes then these changes should be performed upstream.

**Comment by sanette** [comment]
Hi Daniel
thanks for the feedback. Well, it's essentially negative, but at least it's a feeback.
Concerning your points:

1. I see what you mean, but I don't completely agree. If you tried the site, you'd see that once you select a version for some item, the version propagates automaticallly to all other items. Hence, after one first choice, you just need to click on the green button (or links) of other items, nothing else. The choice of version could also be saved in a cookie. (not implemented yet). If you have a better suggestion, please tell.

2. I'm ok to do whatever you prefer, but please guys make up your minds!

Here is my opinion, for what it's worth. Of course, I understand that if the ocaml-manual is the defacto 'package' to store the manuals, then work should be done directly there. However, my approach is not to be conservative and implement experiments fast enough. The only way to test the experiment is to implement them in the real web site. If it doesn't work, we can step back. If it works, we can make it more stable and even prettier, and of course this means pushing it upstream as you recommended.

3. I'm not sure I understand your point.

Concerning your conclusion, to be clear, yes of course I'm interested in structural changes, that's obvious enough. If you decide to simply mirror the existing manual, I'll let you do this, and direct my efforts upstream or elsewhere. But that would be too bad for the end user, imho.

**Comment by dbuenzli** [comment]
> 1. I see what you mean, but I don't completely agree. If you tried the site, you'd see that once you select a version for some item, the version propagates automaticallly to all other items. Hence, after one first choice, you just need to click on the green button (or links) of other items, nothing else. The choice of version could also be saved in a cookie. (not implemented yet). If you have a better suggestion, please tell.

My suggestion is to keep simple things simple (see the end of my previous message). There's no need for menus, JavaScript, cookies and what not for this; plain old and boring html files, hyperlinks and a good URI convention will perfectly do.

> 3. I'm not sure I understand your point.

The point is that you don't want to have structural differences between the manual that is online on ocaml.org and the one you install locally on your machine. This makes it needlessly hard for users to orient themselves or to orient others.

**Comment by sanette** [comment]
> The point is that you don't want to have structural differences between the manual that is online on ocaml.org and the one you install locally on your machine. This makes it needlessly hard for users to orient themselves or to orient others.

There are two different things: one is the manual (tutorials etc) and another one is the automatically generated 'api'. While it might indeed be desirable that the local version of the API matches the online one, I don't see this as crucial for the tutorial part. For the API part, it is not difficult to achieve once people agree on the doc generator (ocamldoc or odoc)

**Comment by dbuenzli** [comment]
> There are two different things: one is the manual (tutorials etc) and another one is the automatically generated 'api'. While it might indeed be desirable that the local version of the API matches the online one, I don't see this as crucial for the tutorial part. 

This is exactly what I see it as very crucial. If I'm reading the manual locally and want to point someone to the section I'm reading I simply want to find the same manual online and hunt the right section there (or even simply hack my local uri manually thanks to an obvious uri structure on `ocaml.org`). 

What I really don't want to is go on ocaml.org, click through menus and have to sift through another version of the manual to find the link I want to communicate.

In fact it's the API that I see as less essential. I hope we can eventually convince upstream to provide a good `odoc` experience for it so that it integrates seamlessly in `odig` generated docsets as the `ocaml` package (see the discussion [here](https://discuss.ocaml.org/t/suggestions-for-ocaml-documentation/4504/20?u=dbuenzli)). This could allow to eventually offload the quite delicate and involved API & manual construction upstream to the mythical `docs.ocaml.org` (a first good version of it being likely [doable](https://discuss.ocaml.org/t/central-opam-documentation-site/3886/17) but rather than trying simple things people giving a shot at this seem obsessed about solving the problem in its full generality, e.g. package versions etc.).

**Comment by sanette** [comment]
> 1. The way to access the manual for a particular version on that screenshot doesn't seem very user friendly. It's unlikely that I want to access these newly created subsets of the manual for a particular version individually, you likely want to consult the manual for a particular version altogether.

Following Daniel's suggestions I have simplified the home page with only one selector for versions:

![Screenshot_20200207_090033](https://user-images.githubusercontent.com/7536536/74011570-82f9a480-4988-11ea-90d9-f190f7dbd64d.png)

**Comment by yawaramin** [comment]
> There's no need for menus, JavaScript, cookies and what not for this; plain old and boring html files, hyperlinks and a good URI convention will perfectly do.

@dbuenzli how would the search box work?

**Comment by dbuenzli** [comment]
> @dbuenzli how would the search box work?

There are other websites specializing in search that will likely do a better job than a custom search box.

**Comment by sanette** [comment]
> @pmetzger @sanette any news on this?

Here is my PR:

https://github.com/ocaml/ocaml.org/pull/1115
